src/pyramid_retry/__init__.py

```diff
@@ -81,22 +81,28 @@ def retry_policy(environ, router):
         # make the original request
         request_ctx = router.request_context(environ)
         request = request_ctx.begin()
-
-        if activate_hook:
-            retry_attempts = activate_hook(request)
-            if retry_attempts is None:
-                retry_attempts = attempts
+        try:
+            if activate_hook:
+                retry_attempts = activate_hook(request)
+                if retry_attempts is None:
+                    retry_attempts = attempts
+                else:
+                    assert retry_attempts > 0
             else:
-                assert retry_attempts > 0
-        else:
-            retry_attempts = attempts
-
-        # if we are supporting multiple attempts then we must make
-        # make the body seekable in order to re-use it across multiple
-        # attempts. make_body_seekable will copy wsgi.input if
-        # necessary, otherwise it will rewind the copy to position zero
-        if retry_attempts != 1:
-            request.make_body_seekable()
+                retry_attempts = attempts
+
+            # if we are supporting multiple attempts then we must make
+            # make the body seekable in order to re-use it across multiple
+            # attempts. make_body_seekable will copy wsgi.input if
+            # necessary, otherwise it will rewind the copy to position zero
+            if retry_attempts != 1:
+                request.make_body_seekable()
+
+        # Catch make_body_seekable (e.g. 408 RequestTimeout)
+        # and activate_hook exceptions and clean up.
+        except Exception:
+            request_ctx.end()
+            raise
```

Comment:
I think this should be a `finally` so you don't have to reraise the exception:
```
        finally:
            request_ctx.end()
```

Reply:
I think the `finally` will be run even if the try doesn't produce an exception. In which case we don't want to call `request_ctx.end()` as it is used later. The whole thing could probably be cleaned up in a context manger, but the initial setup and for loop that keeps pushing/popping requests from the stack makes this tricky.

Reply:
You are right. I wonder whether this should be a bare except to match the behaviour of the finally further down.

Reply:
I initially had a bare except and their tox suite gave a pretty stern flake8 error:

```
lint run-test: commands[0] | flake8 src/pyramid_retry/
src/pyramid_retry/__init__.py:103:9: E722 do not use bare 'except'
        except:
        ^
ERROR: InvocationError for command /Users/keenan/Desktop/projects/github/pyramid_retry/.tox/lint/bin/flake8 src/pyramid_retry/ (exited with code 1)
```

In any case Exception probably covers most use cases? I'm assuming `system-exiting exceptions` will clear the stack for us.
```
exception Exception
All built-in, non-system-exiting exceptions are derived from this class
```

Reply:
You could probably use `except BaseException` to avoid the linter error. GeneratorExit also inherits directly from BaseException, so I think it's worth being comprehensive in error catching here as user code may inadvertently do something strange.

Reply:
Sounds like a good idea. Pushed code and test.

Reply:
Thanks for the review @lrowe.